pep-0xxx.rst

```diff
@@ -0,0 +1,571 @@
+   the standard library: instead, the standard library uses OpenSSL cipher
+   suite strings.
+5. Reporting errors to the caller, currently implemented by the `SSLError`_
+   class in the ``ssl`` module.
+
+While it is technically possible to define (2) in terms of (3), for the sake of
+simplicity it is easier to define these as two separate ABCs. Impelementations
+are of course free to implement the concrete subclasses however they see fit.
+
+Obviously, (4) doesn't require an abstract base class: instead, it requires a
+richer API for configuring supported cipher suites that can be easily updated
+with supported cipher suites for different implementations.
+
+Context
+~~~~~~~
+
+The ``Context`` abstract base class defines an object that allows configuration
+of TLS. It can be thought of as a factory for ``TLSWrappedSocket`` and
+``TLSWrappedBuffer`` objects.
+
+The ``Context`` abstract base class has the following class definition::
+
+    TLSBufferObject = Union[TLSWrappedSocket, TLSWrappedBuffer]
+    ServerNameCallback = Callable[[TLSBufferObject, Optional[str], Context], Any]
+
+    class Context(metaclass=ABCMeta):
+        @abstractmethod
+        def register_certificates(self,
+                                  certificates: str,
+                                  key=None: Optional[str],
+                                  password=None: Optional[Callable[[], Union[AnyStr, bytearray]]]) -> None:
+            """
+            Loads a certificate, a number of intermediate certificates, and the
+            corresponding private key. These certificates will be offered to
+            the remote peer during the handshake if required.
+
+            The ``certificates`` argument must be a bytestring containing the
+            PEM-encoded certificates. The first PEM-encoded certificate must be
+            the leaf certificate. All subsequence certificates will be offered
+            as intermediate additional certificates.
+
+            The ``key`` argument, if present, must contain the PEM-encoded
+            private key associated with the leaf certificate. If not present,
+            the private key will be extracted from ``certificates``.
+
+            The ``password`` argument may be a function to call to get the
+            password for decrypting the private key. It will only be called if
+            the private key is encrypted and a password is necessary. It will
+            be called with no arguments, and it should return a string, bytes,
+            or bytearray. If the return value is a string it will be encoded as
+            UTF-8 before using it to decrypt the key. Alternatively a string,
+            bytes, or bytearray value may be supplied directly as the password
+            argument. It will be ignored if the private key is not encrypted
+            and no password is needed.
+            """
+            pass
+
+        @abstractmethod
+        def set_ciphers(self, ciphers: List[Ciphers]) -> None:
```

Comment:
I have a database of all known cipher suites, their hex ids and aliases for different TLS libs, https://raw.githubusercontent.com/tiran/tlsdb/master/tlsdb.json